### DISCUSSION TITLE
Use units in a more explicit way

### DISCUSSION THREAD

**Discussion by AlexisDrogoul** [gama-platform/gama#37] [Ideas]
Right now, units (like `#km`, `#hour`) are mostly used for computation. While they work quite nicely, they are not explicit in the sense that a variable does not know in which unit it has been expressed. This information however can be important for the user, esp. when displaying values (`10 #m/#s` is more explicit than `0,002777777778`…)

This could be done by adding to attributes/variables a new facet (`unit:`), which would represent their unit. For instance:

`float a <- 100 unit: #km/#h;`

The internal value kept by the variable would be 100, but whenever it would be used in a computation, it would compute its "actual" value (`27,7777…`) and `stringValue(…)` and `serialize(…)` would return `100 #km/#h`. 

The omission of `unit:` would simply signify that the value is to be considered as expressed in the default unit. 

The same would be true if the variable is used in computations:

`float b <- a + 10;` would be interpreted as `100 #km/#h + 10#m/#s` (the latter being the default units of distance and time).
`float b <- a + 10 #km/#h`  would be correctly interpreted.

To do that, we need to change the getValue() and setValue() of numeric variables to take into account their unit (if present): getting the value would return the 'internal' value multiplied by the value of the unit, setting a value would divide it by the unit first, and then assign the result to the 'internal' value.

Any comment ? ( @gama-platform/committers / @gama-platform/core )

**Comment by dphilippon** [comment]
I agree with that, even more knowing that in the GUI it's the actual value which is written.
However, I think that the part with float b and the computation are quite unclear for me as the second writing `float b <- a + 10 #km/#h;` corresponds exactly as what we have right now (which is what we are changing).
 I know that if i do that, i'd be surprise to see the result equals to 136 km per hour... So I would prefer to precise a unit facet (as for float a but this time for float b) instead of writing `10 #km/#h` or to consider the value added with the same unit of the variable a.

**Comment by AlexisDrogoul** [comment]
To make things clear, if I define:

`float a <- 100 unit: #km/#h;`

and if I do:

`float b <- a + 10 #km/#h;`

What I will have **internally** in `b` depends on how it is defined. If it is a temp variable or an attribute without unit, the result would be the same as what we have now (i.e. `30,55555555…`). If it has a unit, then, it would depend on that unit. But **externally**, nothing will be changed.

The goal of this enhancement is (1) to keep the easiness with which we manipulate unitless variables and values; (2) maintain the information about units at the level of agents attributes so that we can display it later.

The only choice we have to make is on how to interpret the operations between values with and without units. I would tend to favor a situation close to the actual one, which is to never make a choice for the user. I.e. if I write:

`a <- a + 10;` we should **not** interpret it as `a <- a + 10#km/#h;` but as `a <- a + 10 #m/#s;` (the default units). It is the responsibility of the modeler to correctly label their values with units (which is what we can already do now, but we lose this information for attributes).

So I dont want to change what we have already (which works quite well), but only to be more explicit when it comes to units in attributes. After that, if someone writes:

`float a <- 100 #km/#h;`, it will work exactly in the same way, but the information about the unit will simply be lost next time we read the value of a.

**Comment by AlexisDrogoul** [comment]
Would be useful also for #1705 in the light of what has been decided regarding the deprecation of `cycle_length` and the use of model units to express the time scaling factor.

**Comment by lesquoyb** [comment]
Is this still something we want to do ?
I also have some questions regarding it:
- what happens when we do operations between units that are not compatible ? ( for example 3#m + 2#L)
  - if we go for a "we don't care if it's compatible or not" strategy then we could lose the unit information silently
  - if we raise an error or something similar for incompatible types it means we will have to extensively define what are the possible conversion (maybe it's easy to do, I didn't really thought it through but maybe not)
- can we come back easily from a number with a unit to a number without ?
- can we have "custom" units ? made by the users, either in gaml or in java plugins
